Build fails on 32-bit architectures

Hi,

Version 1.4 fails on e.g. i386 with:

```
	dune build -j 6 -p eio,eio_main,eio_posix,eio_linux
File "lib_eio_posix/dune", lines 12-15, characters 0-102:
12 | (rule
13 |  (targets config.ml)
14 |  (enabled_if (= %{os_type} "Unix"))
15 |  (action (run ./include/discover.exe)))
(cd _build/default/lib_eio_posix && include/discover.exe)
compiling c program:
 | #include <stdio.h>
 | #include <errno.h>
 | #include <sys/types.h>
 | #include <sys/stat.h>
 | #include <fcntl.h>
 | #include <limits.h>
 | #include <sys/socket.h>
 | 
 | 
 | #ifdef O_DSYNC
 | const char *s0 = "BEGIN-0-true-END";
 | #else
 | const char *s0 = "BEGIN-0-false-END";
 | #endif
 | 
 | 
 | #ifdef O_RESOLVE_BENEATH
 | const char *s1 = "BEGIN-1-true-END";
 | #else
 | const char *s1 = "BEGIN-1-false-END";
 | #endif
 | 
 | 
 | #ifdef O_PATH
 | const char *s2 = "BEGIN-2-true-END";
 | #else
 | const char *s2 = "BEGIN-2-false-END";
 | #endif
 | 
 | 
 | #ifdef MSG_CMSG_CLOEXEC
 | const char *s3 = "BEGIN-3-true-END";
 | #else
 | const char *s3 = "BEGIN-3-false-END";
 | #endif
 | 
run: i686-linux-gnu-gcc -O2 -fno-strict-aliasing -fwrapv -fPIC  -pthread  -D_LARGEFILE64_SOURCE -D_XOPEN_SOURCE=700 -D_DARWIN_C_SOURCE -D_GNU_SOURCE -D_BSD_SOURCE -D__BSD_VISIBLE -I /usr/lib/i386-linux-gnu/ocaml/5.4.1 -o /tmp/build_39230c_dune/ocaml-configuratore6b0ed/c-test-0/test.o -c /tmp/build_39230c_dune/ocaml-configuratore6b0ed/c-test-0/test.c -latomic -lm -lpthread
-> process exited with code 0
-> stdout:
-> stderr:
 | In file included from /usr/include/i386-linux-gnu/bits/libc-header-start.h:33,
 |                  from /usr/include/stdio.h:28,
 |                  from /tmp/build_39230c_dune/ocaml-configuratore6b0ed/c-test-0/test.c:1:
 | /usr/include/features.h:199:3: warning: #warning "_BSD_SOURCE and _SVID_SOURCE are deprecated, use _DEFAULT_SOURCE" [-Wcpp]
 |   199 | # warning "_BSD_SOURCE and _SVID_SOURCE are deprecated, use _DEFAULT_SOURCE"
 |       |   ^~~~~~~
compiling c program:
 | #include <stdio.h>
 | #include <errno.h>
 | #include <sys/types.h>
 | #include <sys/stat.h>
 | #include <fcntl.h>
 | #include <limits.h>
 | #include <sys/socket.h>
 | 
 | 
 | #define DUNE_ABS(x) ((x >= 0)? x: -(x))
 | #define DUNE_D0(x) ('0'+(DUNE_ABS(x)/1         )%10)
 | #define DUNE_D1(x) ('0'+(DUNE_ABS(x)/10        )%10), DUNE_D0(x)
 | #define DUNE_D2(x) ('0'+(DUNE_ABS(x)/100       )%10), DUNE_D1(x)
 | #define DUNE_D3(x) ('0'+(DUNE_ABS(x)/1000      )%10), DUNE_D2(x)
 | #define DUNE_D4(x) ('0'+(DUNE_ABS(x)/10000     )%10), DUNE_D3(x)
 | #define DUNE_D5(x) ('0'+(DUNE_ABS(x)/100000    )%10), DUNE_D4(x)
 | #define DUNE_D6(x) ('0'+(DUNE_ABS(x)/1000000   )%10), DUNE_D5(x)
 | #define DUNE_D7(x) ('0'+(DUNE_ABS(x)/10000000  )%10), DUNE_D6(x)
 | #define DUNE_D8(x) ('0'+(DUNE_ABS(x)/100000000 )%10), DUNE_D7(x)
 | #define DUNE_D9(x) ('0'+(DUNE_ABS(x)/1000000000)%10), DUNE_D8(x)
 | #define DUNE_SIGN(x) ((x >= 0)? '0': '-')
 | 
 | 
 | const char s0[] = {
 |   'B', 'E', 'G', 'I', 'N', '-', '0', '-',
 |   DUNE_SIGN((O_DSYNC)),
 |   DUNE_D9((O_DSYNC)),
 |   '-', 'E', 'N', 'D'
 | };
 | 
 | 
 | const char s1[] = {
 |   'B', 'E', 'G', 'I', 'N', '-', '1', '-',
 |   DUNE_SIGN((O_PATH)),
 |   DUNE_D9((O_PATH)),
 |   '-', 'E', 'N', 'D'
 | };
 | 
 | 
 | const char s2[] = {
 |   'B', 'E', 'G', 'I', 'N', '-', '2', '-',
 |   DUNE_SIGN((MSG_CMSG_CLOEXEC)),
 |   DUNE_D9((MSG_CMSG_CLOEXEC)),
 |   '-', 'E', 'N', 'D'
 | };
 | 
 | 
 | const char s3[] = {
 |   'B', 'E', 'G', 'I', 'N', '-', '3', '-',
 |   DUNE_SIGN((O_RDONLY)),
 |   DUNE_D9((O_RDONLY)),
 |   '-', 'E', 'N', 'D'
 | };
 | 
 | 
 | const char s4[] = {
 |   'B', 'E', 'G', 'I', 'N', '-', '4', '-',
 |   DUNE_SIGN((O_RDWR)),
 |   DUNE_D9((O_RDWR)),
 |   '-', 'E', 'N', 'D'
 | };
 | 
 | 
 | const char s5[] = {
 |   'B', 'E', 'G', 'I', 'N', '-', '5', '-',
 |   DUNE_SIGN((O_WRONLY)),
 |   DUNE_D9((O_WRONLY)),
 |   '-', 'E', 'N', 'D'
 | };
 | 
 | 
 | const char s6[] = {
 |   'B', 'E', 'G', 'I', 'N', '-', '6', '-',
 |   DUNE_SIGN((O_APPEND)),
 |   DUNE_D9((O_APPEND)),
 |   '-', 'E', 'N', 'D'
 | };
 | 
 | 
 | const char s7[] = {
 |   'B', 'E', 'G', 'I', 'N', '-', '7', '-',
 |   DUNE_SIGN((O_CLOEXEC)),
 |   DUNE_D9((O_CLOEXEC)),
 |   '-', 'E', 'N', 'D'
 | };
 | 
 | 
 | const char s8[] = {
 |   'B', 'E', 'G', 'I', 'N', '-', '8', '-',
 |   DUNE_SIGN((O_CREAT)),
 |   DUNE_D9((O_CREAT)),
 |   '-', 'E', 'N', 'D'
 | };
 | 
 | 
 | const char s9[] = {
 |   'B', 'E', 'G', 'I', 'N', '-', '9', '-',
 |   DUNE_SIGN((O_DIRECTORY)),
 |   DUNE_D9((O_DIRECTORY)),
 |   '-', 'E', 'N', 'D'
 | };
 | 
 | 
 | const char s10[] = {
 |   'B', 'E', 'G', 'I', 'N', '-', '1', '0', '-',
 |   DUNE_SIGN((O_EXCL)),
 |   DUNE_D9((O_EXCL)),
 |   '-', 'E', 'N', 'D'
 | };
 | 
 | 
 | const char s11[] = {
 |   'B', 'E', 'G', 'I', 'N', '-', '1', '1', '-',
 |   DUNE_SIGN((O_NOCTTY)),
 |   DUNE_D9((O_NOCTTY)),
 |   '-', 'E', 'N', 'D'
 | };
 | 
 | 
 | const char s12[] = {
 |   'B', 'E', 'G', 'I', 'N', '-', '1', '2', '-',
 |   DUNE_SIGN((O_NOFOLLOW)),
 |   DUNE_D9((O_NOFOLLOW)),
 |   '-', 'E', 'N', 'D'
 | };
 | 
 | 
 | const char s13[] = {
 |   'B', 'E', 'G', 'I', 'N', '-', '1', '3', '-',
 |   DUNE_SIGN((O_NONBLOCK)),
 |   DUNE_D9((O_NONBLOCK)),
 |   '-', 'E', 'N', 'D'
 | };
 | 
 | 
 | const char s14[] = {
 |   'B', 'E', 'G', 'I', 'N', '-', '1', '4', '-',
 |   DUNE_SIGN((O_SYNC)),
 |   DUNE_D9((O_SYNC)),
 |   '-', 'E', 'N', 'D'
 | };
 | 
 | 
 | const char s15[] = {
 |   'B', 'E', 'G', 'I', 'N', '-', '1', '5', '-',
 |   DUNE_SIGN((O_TRUNC)),
 |   DUNE_D9((O_TRUNC)),
 |   '-', 'E', 'N', 'D'
 | };
 | 
 | 
 | const char s16[] = {
 |   'B', 'E', 'G', 'I', 'N', '-', '1', '6', '-',
 |   DUNE_SIGN((AT_FDCWD)),
 |   DUNE_D9((AT_FDCWD)),
 |   '-', 'E', 'N', 'D'
 | };
 | 
 | 
 | const char s17[] = {
 |   'B', 'E', 'G', 'I', 'N', '-', '1', '7', '-',
 |   DUNE_SIGN((AT_SYMLINK_NOFOLLOW)),
 |   DUNE_D9((AT_SYMLINK_NOFOLLOW)),
 |   '-', 'E', 'N', 'D'
 | };
 | 
 | 
 | const char s18[] = {
 |   'B', 'E', 'G', 'I', 'N', '-', '1', '8', '-',
 |   DUNE_SIGN((IOV_MAX)),
 |   DUNE_D9((IOV_MAX)),
 |   '-', 'E', 'N', 'D'
 | };
 | 
run: i686-linux-gnu-gcc -O2 -fno-strict-aliasing -fwrapv -fPIC  -pthread  -D_LARGEFILE64_SOURCE -D_XOPEN_SOURCE=700 -D_DARWIN_C_SOURCE -D_GNU_SOURCE -D_BSD_SOURCE -D__BSD_VISIBLE -I /usr/lib/i386-linux-gnu/ocaml/5.4.1 -o /tmp/build_39230c_dune/ocaml-configuratore6b0ed/c-test-2/test.o -c /tmp/build_39230c_dune/ocaml-configuratore6b0ed/c-test-2/test.c -latomic -lm -lpthread
-> process exited with code 0
-> stdout:
-> stderr:
 | In file included from /usr/include/i386-linux-gnu/bits/libc-header-start.h:33,
 |                  from /usr/include/stdio.h:28,
 |                  from /tmp/build_39230c_dune/ocaml-configuratore6b0ed/c-test-2/test.c:1:
 | /usr/include/features.h:199:3: warning: #warning "_BSD_SOURCE and _SVID_SOURCE are deprecated, use _DEFAULT_SOURCE" [-Wcpp]
 |   199 | # warning "_BSD_SOURCE and _SVID_SOURCE are deprecated, use _DEFAULT_SOURCE"
 |       |   ^~~~~~~
Error: Unable to read variable "MSG_CMSG_CLOEXEC" of type int. Invalid value "01073741824" in /tmp/build_39230c_dune/ocaml-configuratore6b0ed/c-test-2/test.o found
```

--- a/lib_eio_posix/eio_posix_stubs.c
+++ b/lib_eio_posix/eio_posix_stubs.c
@@ -521,11 +521,14 @@ static value get_msghdr_fds(struct msghdr *msg) {
       int n_fds = (cm->cmsg_len - CMSG_LEN(0)) / sizeof(int);
       int i;
       for (i = n_fds - 1; i >= 0; i--) {
-	value fd = Val_int(fds[i]);
-	v_cons = caml_alloc_tuple(2);
-	Store_field(v_cons, 0, fd);
-	Store_field(v_cons, 1, v_list);
-	v_list = v_cons;
+        int fd = fds[i];
+#ifndef MSG_CMSG_CLOEXEC
+        fcntl(fd, F_SETFD, FD_CLOEXEC);         // For macOS
+#endif
+        v_cons = caml_alloc_tuple(2);
+        Store_field(v_cons, 0, Val_int(fd));
+        Store_field(v_cons, 1, v_list);
+        v_list = v_cons;
       }
     }
   }
@@ -564,6 +567,10 @@ CAMLprim value caml_eio_posix_recv_msg(value v_fd, value v_max_fds, value v_bufs
     .msg_controllen = controllen,
   };
   ssize_t r;
+  int flags = Int_val(v_flags);
+#ifdef MSG_CMSG_CLOEXEC
+  flags |= MSG_CMSG_CLOEXEC;
+#endif
 
   memset(cmsg, 0, controllen);
 
@@ -571,7 +578,7 @@ CAMLprim value caml_eio_posix_recv_msg(value v_fd, value v_max_fds, value v_bufs
   msg.msg_iov = iov;
 
   caml_enter_blocking_section();
-  r = recvmsg(Int_val(v_fd), &msg, Int_val(v_flags));
+  r = recvmsg(Int_val(v_fd), &msg, flags);
   caml_leave_blocking_section();
   caml_stat_free_preserving_errno(iov);
   if (r < 0) uerror("recv_msg", Nothing);

--- a/lib_eio_posix/include/discover.ml
+++ b/lib_eio_posix/include/discover.ml
@@ -4,7 +4,6 @@ let optional_flags = [
   "O_DSYNC";
   "O_RESOLVE_BENEATH";
   "O_PATH";
-  "MSG_CMSG_CLOEXEC";
 ]
 
 let () =

--- a/lib_eio_posix/low_level.ml
+++ b/lib_eio_posix/low_level.ml
@@ -124,10 +124,9 @@ let recv_msg fd buf =
 let recv_msg_with_fds ~sw ~max_fds fd buf =
   let addr, got, fds =
     Fd.use_exn "recv_msg" fd @@ fun fd ->
-    let flags = Option.value Config.msg_cmsg_cloexec ~default:0 in
+    let flags = 0 in
     do_nonblocking Read "recv_msg" (fun fd -> eio_recv_msg fd max_fds buf flags) fd
   in
-  if Config.msg_cmsg_cloexec = None then List.iter Unix.set_close_on_exec fds; (* For macos *)
   (addr, got, Eio_unix.Fd.of_unix_list ~sw fds)
 
 external eio_getrandom : Cstruct.buffer -> int -> int -> int = "caml_eio_posix_getrandom"

